### PR TITLE
Create IPuppet interface

### LINKED ISSUE #1592: Create the IPuppet interface
Create the IPuppet interface in support of the agent refactor effort <sup>[1](#594), [2](https://guardicore.atlassian.net/wiki/spaces/RES/pages/2818932869/Infection+Monkey+-+Monkey+Agent+Refactor-+Phase+1+-+PDR)</sup>. Implement a mock IPuppet object that returns canned responses to specific inputs.

- [x] Create the IPuppet interface (0d) @mssalvatore 
- [x] Create a mock Puppet (0d) @mssalvatore

### PR DESCRIPTION
**Pull request by mssalvatore**
# What does this PR do? 

Fixes #1592

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing? 
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [ ] ~Have you successfully tested your changes locally? Elaborate:~
    > Tested by {Running the Monkey locally with relevant config/running Island/...} 
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

**Review by ilija-lazoroski — APPROVED** [review]

File: monkey/infection_monkey/i_puppet.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class IPuppet(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def run_sys_info_collector(self, name: str) -> Dict:
+        """
+        Runs a system info collector
+        :param str name: The name of the system info collector to run
+        :return: A dictionary containing the information collected from the system
+        :rtype: Dict
```
Comment (shreyamalviya) [inline monkey/infection_monkey/i_puppet.py:23]: Is this really needed? We already have type annotations.

File: monkey/infection_monkey/i_puppet.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        Runs a post-breach action (PBA)
+        :param str name: The name of the post-breach action to run
+        :param Dict options: A dictionary containing options that modify the behavior of the PBA
+        """
+
+    @abc.abstractmethod
+    def ping(self, host: str) -> Tuple[bool, Optional[str]]:
+        """
+        Sends a ping (ICMP packet) to a remote host
+        :param str host: The domain name or IP address of a host
```
Comment (shreyamalviya) [inline monkey/infection_monkey/i_puppet.py:38]: `host` right now, in a lot of places we're using it, is an object of type `VictimHost`. Just pointing it out so we're sure this is going to be a string.

File: monkey/infection_monkey/i_puppet.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        Runs a fingerprinter against a remote host
+        :param str name: The name of the fingerprinter to run
+        :param str host: The domain name or IP address of a host
+        :return: A dictionary containing the information collected by the fingerprinter
+        :rtype: Dict
+        """
+
+    @abc.abstractmethod
+    def exploit_host(self, name: str, host: str, options: dict, interrupt: threading.Event) -> bool:
```
Comment (shreyamalviya) [inline monkey/infection_monkey/i_puppet.py:65]: Does `dict` work or do we need `Dict` from `typing`?

**Review by shreyamalviya — COMMENTED** [review]
Some comments (valid in several places).

File: monkey/infection_monkey/i_puppet.py (hunk 1 above)
Comment (mssalvatore) [reply monkey/infection_monkey/i_puppet.py:23]: Yeah. This is used for generating documentation. We're not doing it at the moment, but we should in the future. We don't need to add these kinds of comments for everything, but we really should have them for our interfaces.

File: monkey/infection_monkey/i_puppet.py (hunk 2 above)
Comment (mssalvatore) [reply monkey/infection_monkey/i_puppet.py:38]: Yeah. It's a string.

File: monkey/infection_monkey/i_puppet.py (hunk 3 above)
Comment (mssalvatore) [reply monkey/infection_monkey/i_puppet.py:65]: It seems to work, but I'll change it to `Dict` for consistency.

File: monkey/infection_monkey/i_puppet.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        Sends a ping (ICMP packet) to a remote host
+        :param str host: The domain name or IP address of a host
+        :return: A tuple that contains whether or not the host responded and the host's inferred
+                 operating system
+        :rtype: Tuple[bool, Optional[str]]
+        """
+
+    @abc.abstractmethod
+    def scan_tcp_port(self, host: str, port: int) -> PortScanData:
```
Comment (shreyamalviya) [inline monkey/infection_monkey/i_puppet.py:45]: I believe this needs an `options` argument for `tcp_scan_timeout` and `tcp_scan_interval_ config options.

Comment (mssalvatore) [reply monkey/infection_monkey/i_puppet.py:45]: `tcp_scan_interval` will be handled by the master. I'll add `tcp_scan_timeout`.

**Pull request merged**